services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaController.kt

```diff
@@ -53,14 +53,44 @@ class HankeKayttajaController(
             ]
     )
     @PreAuthorize("@hankeKayttajaAuthorizer.authorizeHankeTunnus(#hankeTunnus, 'VIEW')")
-    fun whoami(@PathVariable hankeTunnus: String): WhoamiResponse {
+    fun whoami(@PathVariable hankeTunnus: String): WhoAmIDto {
         val userId = currentUserId()
 
         val hankeIdentifier = hankeService.findIdentifier(hankeTunnus)!!
         val permission = permissionService.findPermission(hankeIdentifier.id, userId)!!
 
         val hankeKayttaja = hankeKayttajaService.getKayttajaByUserId(hankeIdentifier.id, userId)
-        return WhoamiResponse(hankeKayttaja?.id, permission.kayttooikeustasoEntity)
+        return WhoAmIDto(hankeKayttaja?.id, permission.kayttooikeustasoEntity)
+    }
+
+    @GetMapping("hankkeet/my-permissions")
+    @Operation(
+        summary = "Get your permissions for your own projects",
+        description = "Returns a map of current users Hanke identifiers and respective permissions."
+    )
+    @ApiResponses(
+        value =
+            [
+                ApiResponse(
+                    description = "Permissions grouped by hankeTunnus.",
+                    responseCode = "200",
+                ),
+                ApiResponse(description = "Access denied.", responseCode = "401")
+            ]
+    )
+    fun whoAmIByHanke(): Map<String, WhoAmIDto> {
+        val whoAmIAtHanke: Map<Int, WhoAmIDto> =
+            hankeKayttajaService.whoAmIByHankeId(userId = currentUserId())
+        val hankeIds = whoAmIAtHanke.keys
+
+        if (hankeIds.isEmpty()) return emptyMap()
+
+        return hankeService
+            .findIdentifiers(hankeIds)
+            .mapNotNull { (hankeId, hankeTunnus) ->
```

Comment:
I think a lot of the complexities in combining the information from different tables could be avoided by using a custom query to get the information all at once. Maybe something like (in PermissionRepository):
```kotlin
    @Query(
        "select new fi.hel.haitaton.hanke.permissions.HankePermission(" +
            "  h.hankeTunnus, hk.id, pe.kayttooikeustasoEntity" +
            ") from PermissionEntity pe " +
            "inner join HankeEntity h on pe.hankeId = h.id " +
            "left join HankeKayttajaEntity hk on hk.permission.id = pe.id " +
            "where pe.userId = :userId"
    )
    fun userPermissions(userId: String): List<HankePermission>
```

Reply:
HankePermission would also be the domain object needed to avoid using the DTO on the service level.

Maybe the kayttooikeustasoEntity could be unpacked in the query as well:
```
    @Query(
        "select new fi.hel.haitaton.hanke.permissions.HankePermission(" +
            "  h.hankeTunnus," +
            "  hk.id," +
            "  pe.kayttooikeustasoEntity.kayttooikeustaso," +
            "  pe.kayttooikeustasoEntity.permissionCode" +
            ") from PermissionEntity pe " +
            "inner join HankeEntity h on pe.hankeId = h.id " +
            "left join HankeKayttajaEntity hk on hk.permission.id = pe.id " +
            "where pe.userId = :userId"
    )
    fun userPermissions(userId: String): List<HankePermission>
```

Reply:
I agree there is some complexity to this. However, I generally consider a good practice to keep repositories focused on their respective entities. Additionally, that query is not a simple one either. I'll try to think of a way to solve this.

Reply:
> However, I generally consider a good practice to keep repositories focused on their respective entities.
Wouldn't this mean that all queries returning custom objects are forbidden? This is a big limitation.

Reply:
No, it is just good practice. 

I changed the implementation as suggested as I did not see any other way around this. Added an index for permissions.userid also.

Reply:
Well spotted on the index 👍